runtime/core/bin/decoder_main.cc

```diff
@@ -129,13 +111,51 @@ int main(int argc, char *argv[]) {
         buffer << "candidate " << r.score << " " << r.sentence << std::endl;
       }
     }
+    total_waves_dur_ += wave_dur;
+    total_decode_time_ += decode_time;
+    has_done_+=1;
+    mutex_.unlock();
+}
 
-    total_waves_dur += wave_dur;
-    total_decode_time += decode_time;
+int main(int argc, char *argv[]) {
+  gflags::ParseCommandLineFlags(&argc, &argv, false);
+  google::InitGoogleLogging(argv[0]);
+
+  decode_config_ = wenet::InitDecodeOptionsFromFlags();
+  feature_config_ = wenet::InitFeaturePipelineConfigFromFlags();
+  decode_resource_ = wenet::InitDecodeResourceFromFlags();
+
+  if (FLAGS_wav_path.empty() && FLAGS_wav_scp.empty()) {
+    LOG(FATAL) << "Please provide the wave path or the wav scp.";
+  }
+  std::vector<std::pair<std::string, std::string>> waves;
+  if (!FLAGS_wav_path.empty()) {
+    waves.emplace_back(make_pair("test", FLAGS_wav_path));
+  } else {
+    std::ifstream wav_scp(FLAGS_wav_scp);
+    std::string line;
+    while (getline(wav_scp, line)) {
+      std::vector<std::string> strs;
+      wenet::SplitString(line, &strs);
+      CHECK_GE(strs.size(), 2);
+      waves.emplace_back(make_pair(strs[0], strs[1]));
+    }
   }
-  LOG(INFO) << "Total: decoded " << total_waves_dur << "ms audio taken "
-            << total_decode_time << "ms.";
+
+  if (!FLAGS_result.empty()) {
+    result_.open(FLAGS_result, std::ios::out);
+  }
+
+  ThreadPool pool(FLAGS_thread_num);
+  for (auto &wav : waves) {
+    pool.enqueue(decode, wav);
+  }
+
+  while (has_done_ != waves.size()) {}
+
```

Comment:
You can just use the following method to shutdown the thread pool.
``` c++
{
    ThreadPool pool(FLAGS_thread_num);
    for (auto &wav : waves) {
      pool.enqueue(decode, wav);
    }
}
```

Sometimes some decoding fails, then the program will be blocked forever in 

``` c++
while (has_done_ != waves.size()) {}
```